### PR TITLE
Complete grant CLI policy interactons and probationary period enforcement

### LINKED ISSUE #2353: Disclaimer for policy probationary period

### PR DESCRIPTION
**Pull request by KPrasch**
Superceeds #2431 and scraps some of #2115  

Closes #2353

### CODE REVIEW

File: nucypher/cli/painting/help.py [hunk 1]
```diff
@@ -74,6 +75,38 @@ def paint_new_installation_help(emitter, new_configuration):
...
     adjective = 'an' if character_name_starts_with_vowel else 'a'
     suggested_command = f'nucypher {character_name} run'
-    how_to_run_message = f"\nTo run {adjective} {character_name.capitalize()} node from the default configuration filepath run: \n\n'{suggested_command}'\n"
+    how_to_run_message = f"\nTo run {adjective} {character_name.capitalize()} node from the default configuration " \
+                         f"filepath run: \n\n'{suggested_command}'\n"
 
     emitter.echo(how_to_run_message.format(suggested_command), color='green')
+
+
+def paint_probationary_period_disclaimer(emitter):
```
Comment (vepkenez) [inline nucypher/cli/painting/help.py:84]: The fact that we have this method called "paint_probationary_period_disclaimer" is kinda crazy.

**Review by vepkenez — APPROVED** [review]

File: nucypher/cli/commands/alice.py [hunk 2]
```diff
@@ -447,13 +461,57 @@ def grant(general_config,
...
+            n = click.prompt('Enter threshold (M)', type=click.IntRange(1, n+1))
+
+    # Policy Value
+    if not ALICE.federated_only and (not (bool(value) or bool(rate))):
         rate = ALICE.default_rate  # TODO #1709
-        click.confirm(f"Confirm default rate {rate}?", abort=True)
+        if not force:
+            use_default = click.confirm(f"Confirm default rate {rate}?", default=True)
+            if not use_default:
+                rate = click.prompt('Enter rate per period in Wei', type=WEI)
```
Comment (cygnusv) [inline nucypher/cli/commands/alice.py:512]: Current rate range is in the GWEI order of magnitude, so better to use the `GWEI` type.

File: newsfragments/2431.misc.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Introduces a probationary period for policy creation in the network, until 2021-01-31 23:59:59 UTC.
```
Comment (cygnusv) [inline newsfragments/2431.misc.rst:1]: Update the newsfragment to the final date

**Review by cygnusv — COMMENTED** [review]
I'll have an additional look within the next 2 hours, but your part looks very good @KPrasch . Thanks for taking time to adopt this PR!

File: nucypher/cli/commands/alice.py (hunk 2 above)
Comment (KPrasch) [reply nucypher/cli/commands/alice.py:512]: :heavy_check_mark:

File: newsfragments/2431.misc.rst (hunk 3 above)
Comment (KPrasch) [reply newsfragments/2431.misc.rst:1]: :heavy_check_mark:

File: nucypher/cli/painting/help.py [hunk 4]
```diff
@@ -74,6 +75,38 @@ def paint_new_installation_help(emitter, new_configuration):
...
+    width = 60
+    import textwrap
+    disclaimer_title = " DISCLAIMER ".center(width, "=")
+    paragraph = f"""
+Some areas of the NuCypher network are still under active development;
+as a consequence, we have established a probationary period for policies in the network.
+We currently recommend not to create data sharing policies with a duration beyond {END_OF_POLICIES_PROBATIONARY_PERIOD}.
```
Comment (derekpierre) [inline nucypher/cli/painting/help.py:91]:
```suggestion
We currently prevent the creation of sharing policies with durations beyond {END_OF_POLICIES_PROBATIONARY_PERIOD}.
```

**Review by derekpierre — CHANGES_REQUESTED** [review]

File: nucypher/characters/control/specifications/alice.py [hunk 5]
```diff
@@ -96,7 +97,7 @@ class GrantPolicy(PolicyBaseSchema):
 
     label = fields.Label(
         load_only=True, required=True,
-        click=options.option_label(required=True))
+        click=options.option_label(required=False))
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/characters/control/specifications/alice.py:100]: Great question :-)

Basically the expected behavior here is that the click option itself is optional to allow for interactive selection and forming a request to the character controller on the layer below with required required fields.

File: nucypher/cli/commands/alice.py [hunk 6]
```diff
@@ -447,13 +461,57 @@ def grant(general_config,
...
+
+    # Policy Threshold and Shares
+    if not n:
+        n = ALICE.n
+        if not force and not click.confirm(f'Use default value for N ({n})?', default=True):
+            n = click.prompt('Enter total number of shares (N)', type=click.INT)
+    if not m:
+        m = ALICE.m
+        if not force and not click.confirm(f'Use default value for M ({m})?', default=True):
+            n = click.prompt('Enter threshold (M)', type=click.IntRange(1, n+1))
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/cli/commands/alice.py:504]: Good catch!

[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/cli/commands/alice.py:504]: :heavy_check_mark: for reference https://click.palletsprojects.com/en/7.x/options/#range-options

File: newsfragments/2460.feature.rst [hunk 7]
```diff
@@ -0,0 +1 @@
+Complete interactive collection of policy parameters via alice grant CLI
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply newsfragments/2460.feature.rst:1]: :heavy_check_mark:

File: nucypher/cli/actions/confirm.py [hunk 8]
```diff
@@ -144,3 +145,10 @@ def verify_upgrade_details(blockchain: Union[BlockchainDeployerInterface, Blockc
     click.confirm(CONFIRM_VERSIONED_UPGRADE.format(contract_name=deployer.contract_name,
                                                    old_version=old_contract.version,
                                                    new_version=new_version), abort=True)
+
+
+def confirm_staged_grant(emitter, grant_request: Dict) -> None:
+    emitter.echo("Successfully staged grant.  Please review the details:\n", color='green')
+    table = ([field, value] for field, value in grant_request.items())
+    emitter.echo(tabulate(table, tablefmt="simple"))
+    click.confirm('\nConfirm granting access policy and sign transaction?', abort=True)
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/cli/actions/confirm.py:154]: :heavy_check_mark: - reworded

File: nucypher/cli/commands/alice.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (KPrasch) [reply nucypher/cli/commands/alice.py:504]: :heavy_check_mark:

File: nucypher/cli/painting/help.py (hunk 4 above)
Comment (KPrasch) [reply nucypher/cli/painting/help.py:91]: :heavy_check_mark: - reworked

File: nucypher/cli/commands/alice.py [hunk 9]
```diff
@@ -509,7 +509,7 @@ def grant(general_config,
         if not force:
             use_default = click.confirm(f"Confirm default rate {rate}?", default=True)
             if not use_default:
-                rate = click.prompt('Enter rate per period in Wei', type=WEI)
+                rate = click.prompt('Enter rate per period in gwei', type=GWEI)
```
Comment (cygnusv) [inline nucypher/cli/commands/alice.py:512]: We still have to convert it to wei internally. Gwei is just for user facing experience.

**Review by cygnusv — APPROVED** [review]
Apart from the gwei conversion comment, all good from my side. I half-authored this PR so I'm probably not the best to review it but there's that 👍

File: nucypher/cli/commands/alice.py (hunk 9 above)
Comment (KPrasch) [reply nucypher/cli/commands/alice.py:512]: Ah, yes of course - thanks!

Comment (KPrasch) [reply nucypher/cli/commands/alice.py]: :heavy_check_mark:

**Pull request merged**